src/GraphTopology.cpp

```diff
@@ -13,469 +13,301 @@
 
-    aggregate_edge_int.Swap(tmp);
+    Init(vertex_edge, partition, edge_edge, edge_true_edge);
 }
 
-// TODO: allow aggregate to be shared by more than one processor
-GraphTopology::GraphTopology(
-    mfem::SparseMatrix& vertex_edge,
-    const mfem::HypreParMatrix& edge_d_td,
-    const mfem::Array<int>& partition,
-    const mfem::SparseMatrix* edge_boundaryattr)
-    : edge_d_td_(edge_d_td),
-      comm_(edge_d_td.GetComm())
+void GraphTopology::Init(const SparseMatrix& vertex_edge,
+                         const std::vector<int>& partition,
+                         const ParMatrix& edge_edge,
+                         const ParMatrix& edge_true_edge)
 {
-    MPI_Comm_size(comm_, &num_procs_);
-    MPI_Comm_rank(comm_, &myid_);
+    MPI_Comm comm = edge_true_edge.GetComm();
 
-    unique_ptr<mfem::HypreParMatrix> edge_td_d( edge_d_td_.Transpose() );
+    agg_vertex_local_ = MakeAggVertex(partition);
 
-    edge_d_td_d_.reset( ParMult(&edge_d_td_, edge_td_d.get()) );
+    SparseMatrix agg_edge_ext = agg_vertex_local_.Mult(vertex_edge);
+    agg_edge_ext.SortIndices();
 
-    int nvertices = vertex_edge.Height();
-    int nedges = vertex_edge.Width();
-    int nAggs = partition.Max() + 1;
+    agg_edge_local_ = RestrictInterior(agg_edge_ext);
 
-    // generate the 'start' array (not true dof)
-    mfem::Array<HYPRE_Int>* start[3] = {&vertex_start_, &edge_start_,
-                                        &aggregate_start_
-                                       };
-    HYPRE_Int nloc[3] = {nvertices, nedges, nAggs};
-    GenerateOffsets(comm_, 3, nloc, start);
+    SparseMatrix edge_ext_agg = agg_edge_ext.Transpose();
 
-    // Construct the relation table aggregate_vertex from partition
+    int num_vertices = vertex_edge.Rows();
+    int num_edges = vertex_edge.Cols();
+    int num_aggs = agg_edge_local_.Rows();
 
-    mfem::SparseMatrix tmp = PartitionToMatrix(partition, nAggs);
-    Agg_vertex_.Swap(tmp);
+    auto starts = parlinalgcpp::GenerateOffsets(comm, {num_vertices, num_edges, num_aggs});
+    const auto& vertex_starts = starts[0];
+    const auto& edge_starts = starts[1];
+    const auto& agg_starts = starts[2];
 
-    mfem::SparseMatrix aggregate_edge(smoothg::Mult(Agg_vertex_,
-                                                    vertex_edge));
+    ParMatrix edge_agg_d(comm, edge_starts, agg_starts, std::move(edge_ext_agg));
+    ParMatrix agg_edge_d = edge_agg_d.Transpose();
 
-    // Need to sort the edge indices to prevent index problem in face_edge
-    aggregate_edge.SortColumnIndices();
+    ParMatrix edge_agg_ext = edge_edge.Mult(edge_agg_d);
+    ParMatrix agg_agg = agg_edge_d.Mult(edge_agg_ext);
 
-    AggregateEdge2AggregateEdgeInt(aggregate_edge, Agg_edge_);
-    mfem::SparseMatrix edge_aggregate(smoothg::Transpose(aggregate_edge));
+    agg_edge_ext = 1.0;
+    SparseMatrix face_agg_int = MakeFaceAggInt(agg_agg);
+    SparseMatrix face_edge_ext = face_agg_int.Mult(agg_edge_ext);
```

Comment:
I prefer to call `face_agg_int` either `intface_agg` or using the convention in the new code `face_int_agg` because int here refers to interior face.

And for `face_edge_ext`, I guess it comes from `face_agg * agg_edge_ext` and the middle is eliminated. There are other places using this naming convention. What is the reason you dont like the original name? The original name in the old code is `intface_Agg_edge` because the algorithm comes from Sec 2.4 of Panayot's paper (doi:10.1002/nla.300), and in that paper, the name is "AEface_AE_face". I think it is easier for other people to follow if we use some what equivalent names (say `face_agg_edge`).

Reply:
No specific reason for the names,  I'll modify these to make them consistent/more clear.